hist/hist/inc/THnChain.h

```diff
@@ -0,0 +1,266 @@
+// @(#)root/hist:$Id$
+// Author: Benjamin Bannier, August 2016
+
+/*************************************************************************
+ * Copyright (C) 1995-2016, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_THnChain
+#define ROOT_THnChain
+
+#include <TAxis.h>
+#include <TFile.h>
+#include <TH1.h>
+#include <TH2.h>
+#include <TH3.h>
+#include <THnBase.h>
+#include <THnSparse.h>
+#include <TObjArray.h>
+#include <TObjString.h>
+#include <TString.h>
+
+/** \class THnChain
+A class to chain together multiple histograms.
+
+This class allows to chain together any `THnBase`-derived (`THn` or `THnSparse`)
+histograms from multiple files. Operations on the axes and projections are
+supported. The intent is to allow convenient merging merging of projections
+of high-dimensional histograms.
+
+\code{.cpp}
+// `file1.root` and `file2.root` contain a `THnSparse` named `hsparse`.
+THnChain hs("hsparse");
+hs.AddFile("file1.root");
+hs.AddFile("file2.root");
+
+// Project out axis 0, integrate over other axes.
+TH1* h0 = hs.Projection(0);
+
+// Project out axis 0, integrate over other axes in their active ranges.
+hs.GetAxis(1)->SetRangeUser(0, 0.1); // select a subrange
+TH1* h0 = hs.Projection(0);
+\endcode
+
+*/
+
+class THnChain : public TObject
+{
+ public:
+   /// Default constructor.
+   ///
+   /// \param name name of the histogram to work on
+   THnChain(const char* name) : fName(name) { fFiles.SetOwner(kTRUE); }
+
+   /// Add a new file to this chain.
+   ///
+   /// \param fileName path of the file to add
+   void AddFile(const char* fileName)
+   {
```

Comment:
Can this move into the source? In general only implementations of functions that will likely be inlined and are perf critical should reside in the header...